### ISSUE TITLE
Trailing slash removed from URL even when required

### ISSUE THREAD

**Issue by sw-clough** [TehShrike/sausage-router#6]
Problem is described here: https://github.com/TehShrike/abstract-state-router/issues/130

The trailing slash is currently being removed here: https://github.com/TehShrike/sausage-router/blob/2176a3a6264e58a6565bcc2ff0d0f809f7a8e622/index.js#L41

I propose changing that function to something like the following;

```javascript
function joinPath(first, second) {
    parts = [
        "", // empty string ensures initial slash
        ...first.split("/").filter(String), // trim all slashes from first path
        ...second.split("/").filter(String), // trim all slashes from second path
        ...(second.slice(-1) === "/" ? [""] : []) // empty string for optional trailing slash
    ]
    return parts.join("/")
}
```

**Comment by TehShrike** [comment]
You're right, trailing slashes should not be trimmed from the second path.

That array-joining method looks a little too clever for my tastes – I'd rather drop in another `trimLeadingSlash` function and use that on the second argument.

If you open a PR with a failing unit test, I can publish the fix soon.

**Comment by sw-clough** [comment]
I'll see what I can do.

Yes I may have gotten a little clever with my solution. However, I don't like using regex for such a simple task, so perhaps the new implementation should just use regular string functions like indexof and slice?

**Comment by TehShrike** [comment]
I'm very comfortable with regular expressions, so I don't mind the current method, since it only requires a single pass over the string, but I would be fine with an implementation that used `charAt` to check the first and last characters in the string.

**Comment by sw-clough** [comment]
Performance is my concern. Regex is slow and should be avoided for such a simple task. Example: https://jsperf.com/fastest-way-to-remove-last-character-in-javascript/7

**Comment by TehShrike** [comment]
Readability/simplicity is my main concern, though I would prefer a method that only requires iterating over the string once at most.  The regex does that, but `charAt`/square brackets + `substring` would work fine too.
